sdks/java/harness/src/test/java/org/apache/beam/fn/harness/SplittablePairWithRestrictionDoFnRunnerTest.java

```diff
@@ -0,0 +1,805 @@
+      }
+
+      @ProcessElement
+      public ProcessContinuation processElement(
+          ProcessContext context,
+          RestrictionTracker<OffsetRange, Long> tracker,
+          ManualWatermarkEstimator<Instant> watermarkEstimator)
+          throws Exception {
+        long checkpointUpperBound = CHECKPOINT_UPPER_BOUND;
+        long position = tracker.currentRestriction().getFrom();
+        boolean claimStatus = true;
+        while (!shouldAbortProcessing()) {
+          claimStatus = tracker.tryClaim(position);
+          if (!claimStatus) {
+            break;
+          } else if (position == SPLIT_ELEMENT) {
+            splitElementProcessed();
+            waitForTrySplitPerformed();
+          }
+          context.outputWithTimestamp(
+              context.element() + ":" + position,
+              GlobalWindow.TIMESTAMP_MIN_VALUE.plus(Duration.millis(position)));
+          watermarkEstimator.setWatermark(
+              GlobalWindow.TIMESTAMP_MIN_VALUE.plus(Duration.millis(position)));
+          position += 1L;
+          if (position == checkpointUpperBound) {
+            break;
+          }
+        }
+        if (!claimStatus) {
+          return ProcessContinuation.stop();
+        } else {
+          return ProcessContinuation.resume().withResumeDelay(Duration.millis(54321L));
+        }
+      }
+
+      @GetInitialRestriction
+      public OffsetRange restriction(@Element String element) {
+        return new OffsetRange(0, Integer.parseInt(element));
+      }
+
+      @NewTracker
+      public RestrictionTracker<OffsetRange, Long> newTracker(
+          @Restriction OffsetRange restriction) {
+        return new OffsetRangeTracker(restriction);
+      }
+
+      @SplitRestriction
+      public void splitRange(@Restriction OffsetRange range, OutputReceiver<OffsetRange> receiver) {
+        receiver.output(new OffsetRange(range.getFrom(), (range.getFrom() + range.getTo()) / 2));
+        receiver.output(new OffsetRange((range.getFrom() + range.getTo()) / 2, range.getTo()));
+      }
+
+      @TruncateRestriction
+      public TruncateResult<OffsetRange> truncateRestriction(@Restriction OffsetRange range)
+          throws Exception {
+        return TruncateResult.of(new OffsetRange(range.getFrom(), range.getTo() / 2));
+      }
+
+      @GetInitialWatermarkEstimatorState
+      public Instant getInitialWatermarkEstimatorState() {
+        return GlobalWindow.TIMESTAMP_MIN_VALUE.plus(Duration.millis(1));
+      }
+
+      @NewWatermarkEstimator
+      public WatermarkEstimators.Manual newWatermarkEstimator(
+          @WatermarkEstimatorState Instant watermark) {
+        return new WatermarkEstimators.Manual(watermark);
+      }
+    }
+
+    /**
+     * A window observing variant of {@link NonWindowObservingTestSplittableDoFn} which uses the
+     * side inputs to choose the checkpoint upper bound.
+     */
+    static class WindowObservingTestSplittableDoFn extends NonWindowObservingTestSplittableDoFn {
```

Comment:
There are static classes copy-pasted from FnApiDoFnRunnerTest, contributing to most of the positive diff. Would it be preferable to simply use the one in FnApiDoFnRunnerTest or moved to a test fixture class?

Reply:
Good idea. The classes are pretty complex and not documented, so they are probably not very general test fixtures. But still, yea, I moved to top level and shared them between the two test classes.